### PR TITLE
[RW-1278][RW-1280][Risk=no] Blrubenstein/domain counts

### PR DESCRIPTION
**Pull request by blrubenstein**
This adds real domain counts, only disables tabs that haven't loaded yet, and adds filtering by vocabulary. We make a new call on filtering in order to properly support pagination once that is added.

### CODE REVIEW

File: ui/src/app/views/concept-homepage/component.html [hunk 1]
```diff
@@ -48,16 +48,21 @@
...
-        <div *ngFor="let domain of conceptDomainList" class="chart-selector-tab">
-          <button (click)="selectDomain(domain.domain)" class="domain-selector-button btn btn-link nav-link nav-item" [class.active]="domain.domain === selectedDomain" [disabled]="searchLoading">
+        <div *ngFor="let domain of conceptDomainCounts" class="chart-selector-tab">
+          <button (click)="selectDomain(domain.domain)" class="domain-selector-button btn btn-link nav-link nav-item" [class.active]="domain.domain === selectedDomain" [disabled]="searchLoading || !completedDomainSearches.includes(domain.domain)">
             <div class="chart-selector-header">
               {{domain.name}}
             </div>
-            <!-- TODO (RW-1277): Make this the counts returned from search. -->
-            <div>{{domain.allConceptCount}}</div>
+            <div>{{domain.concept_count}}</div>
```
Comment (calbach) [inline ui/src/app/views/concept-homepage/component.html]: Hmm, we should fix this - I suspect the underscore was an accident @danqrodney ?

File: ui/src/app/views/concept-homepage/component.html [hunk 2]
```diff
@@ -48,16 +48,21 @@
...
           </button>
         </div>
       </div>
+      <div class="filter-options">
+        Filter by vocabulary:
+        <clr-checkbox class="filter-checkbox" *ngFor="let vocabulary of vocabularies" (click)="filterList()" [(clrChecked)]="vocabulary.selected">
```
Comment (calbach) [inline ui/src/app/views/concept-homepage/component.html]: Huh, what's the significance of `[(clrChecked)]` ? vs just `[]` or just `()`

File: ui/src/app/views/concept-homepage/component.ts [hunk 3]
```diff
@@ -46,13 +54,23 @@ export class ConceptHomepageComponent implements OnInit {
...
+  completedDomainSearches: Array<Domain> = [];
+
+  vocabularies: Array<VocabularyCountSelected> = [];
+
   wsNamespace: string;
   wsId: string;
 
+  // For some reason clr checkboxes trigger click events twice on click. This
+  // is a workaround to not allow multiple filter events to get triggered.
+  blockMultipleSearchFromFilter = true;
```
Comment (calbach) [inline ui/src/app/views/concept-homepage/component.ts]: Have you tried `e.stopPropagation()` in your click callback??

File: ui/src/app/views/concept-homepage/component.ts [hunk 4]
```diff
@@ -105,36 +129,51 @@ export class ConceptHomepageComponent implements OnInit {
...
 
-  setConcepts() {
-    this.concepts = this.conceptsCache.find(
-      conceptDomain => conceptDomain.domain === this.selectedDomain).items;
+  setConceptsAndVocabularies() {
+    const cacheItem = this.conceptsCache.find(
+      conceptDomain => conceptDomain.domain === this.selectedDomain);
+    this.concepts = cacheItem.items;
+    this.vocabularies = [];
+    cacheItem.vocabularyList.forEach((vocabulary) => {
```
Comment (calbach) [inline ui/src/app/views/concept-homepage/component.ts]: opt: map instead of forEach `this.vocatularies = ...map()`

File: ui/src/app/views/concept-homepage/component.ts [hunk 5]
```diff
@@ -144,4 +183,33 @@ export class ConceptHomepageComponent implements OnInit {
   get searchTermNotLongEnough() {
     return this.searchTerm === undefined || this.searchTerm.length < 3;
   }
+
```
Comment (calbach) [inline ui/src/app/views/concept-homepage/component.ts]: nit: excessive number of newlines

File: ui/src/app/views/concept-homepage/component.ts [hunk 6]
```diff
@@ -144,4 +183,33 @@ export class ConceptHomepageComponent implements OnInit {
...
+    }
+    this.searchLoading = true;
+    const request = {
+      query: this.currentSearchString,
+      standardConceptFilter: standardConceptFilter,
+      domain: this.selectedDomain,
+      vocabularyIds: this.vocabularies
+        .filter(vocabulary => vocabulary.selected).map(vocabulary => vocabulary.vocabularyId),
+    };
+    this.conceptsService.searchConcepts(this.wsNamespace, this.wsId, request)
```
Comment (calbach) [inline ui/src/app/views/concept-homepage/component.ts]: Error callback?

File: ui/src/app/views/concept-homepage/component.ts [hunk 7]
```diff
@@ -105,36 +129,51 @@ export class ConceptHomepageComponent implements OnInit {
...
-        domain: conceptDomain.domain
+        domain: conceptDomain.domain,
+        includeDomainCounts: activeTabSearch,
+        includeVocabularyCounts: true
       };
       this.conceptsService.searchConcepts(this.wsNamespace, this.wsId, request)
         .subscribe((response) => {
           numCalls += 1;
-          numCallsSubject.next(numCalls);
+          numCallsSubject.next(conceptDomain.domain);
```
Comment (calbach) [inline ui/src/app/views/concept-homepage/component.ts]: Not really part of the PR but I'm not sure I'm following this `numCallsSubject` approach, I assume it has to do with tracking multiple parallel concept requests. Did you consider using `forkJoin`?

File: ui/src/app/views/concept-homepage/component.html (hunk 1 above)
Comment (blrubenstein) [reply ui/src/app/views/concept-homepage/component.html]: It was, I have already told Dan and he is fixing it. If he gets a fix in before this goes in, I will rebase and change, otherwise I shall send him the two lines to change.

File: ui/src/app/views/concept-homepage/component.html (hunk 2 above)
Comment (blrubenstein) [reply ui/src/app/views/concept-homepage/component.html]: Ooooh! Teaching time, cause I just learned this recently-ish! In angular `[]` is an input binding, and `()` is an output binding. So `[(clrChecked)]` is a two way binding to the variable, such that it defaults to the state passed in, and then changes to the state of the checkbox component are propegated back out to the model variable. Similar to why we do `[(ngModel)]` vs `[inputVariable]` or `(outputHandler)`

File: ui/src/app/views/concept-homepage/component.ts (hunk 3 above)
Comment (blrubenstein) [reply ui/src/app/views/concept-homepage/component.ts]: Will try. Let me look into that.

File: ui/src/app/views/concept-homepage/component.ts (hunk 7 above)
Comment (blrubenstein) [reply ui/src/app/views/concept-homepage/component.ts]: It is handling multiple concept requests. I think `forkJoin` would have worked before, but now it is changed to enable tabs as the data comes back, so I am not blocking on all of them. I can add a comment if you would like?

File: ui/src/app/views/concept-homepage/component.ts (hunk 6 above)
Comment (blrubenstein) [reply ui/src/app/views/concept-homepage/component.ts]: Haven't implemented errors here yet, but I can do that.

File: ui/src/app/views/concept-homepage/component.ts (hunk 3 above)
Comment (blrubenstein) [reply ui/src/app/views/concept-homepage/component.ts]: Tried it, it blocked changing the values too. I think Clarity needs the propagation, but for some reason that triggers multiple events... :(

File: ui/src/app/views/concept-homepage/component.html (hunk 1 above)
Comment (danqrodney) [reply ui/src/app/views/concept-homepage/component.html]: Yeah, I'm fixing this. I'll update the PR to also include the corresponding UI change.

Comment (blrubenstein) [reply ui/src/app/views/concept-homepage/component.html]: There's also a test file that will need to change. (But only when this goes in) I can send you a list of lines of code after this goes in.
